File: src/example/basicGossip/protocols/BasicGossip.java
package example.basicGossip.protocols;

import java.util.Random;

import peersim.cdsim.CDProtocol;
import peersim.config.Configuration;
import peersim.config.FastConfig;
import peersim.core.Linkable;
import peersim.core.Node;
import peersim.edsim.EDProtocol;
import peersim.transport.Transport;
import peersim.vector.SingleValueHolder;
import example.basicGossip.Info;
import example.basicGossip.Usernode;

/**
 * Event driven version of epidemic averaging.
 */
public class BasicGossip extends SingleValueHolder implements CDProtocol,
		EDProtocol {

	public static int fanout;
	public static int networkSize;
	public static int cycles;
	protected int info;

	// --------------------------------------------------------------------------
	// Initialization
	// --------------------------------------------------------------------------

	/**
	 * @param prefix
	 *            string prefix for config properties
	 */
	public BasicGossip(String prefix) {
		super(prefix);
		networkSize = (Configuration.getInt("network.size", 1));
		cycles = (Configuration.getInt("CYCLES", 1));
		fanout = (int) Math.log(networkSize) + 1;
	}

	// --------------------------------------------------------------------------
	// methods
	// --------------------------------------------------------------------------

	/**
	 * This is the standard method the define periodic activity. The frequency
	 * of execution of this method is defined by a
	 * {@link peersim.edsim.CDScheduler} component in the configuration.
	 */
	public void nextCycle(Node node, int pid) {
		// System.out.println("PID " + pid);
		// streamer
		if (node.getIndex() == 0) {
			System.out.println("streamer generating " + info);
			Random r = new Random();
			for (int i = 0; i < fanout; i++) {

				Linkable linkable = (Linkable) node.getProtocol(FastConfig
						.getLinkable(pid));
				if (linkable.degree() > 0) {
					Node peern = linkable.getNeighbor(r.nextInt(linkable
							.degree()));

					// System.out.println("streamer sending to " +
					// peern.getIndex());

					// XXX quick and dirty handling of failures
					// (message would be lost anyway, we save time)
					if (!peern.isUp())
						return;
					((Transport) node.getProtocol(FastConfig.getTransport(pid)))
							.send(node, peern, new Info(info, (Usernode) node,
									0), pid);
				}

			}

			info++;
		}

	}

	public void processEvent(Node node, int pid, Object event) {
	}
}

File: src/example/basicGossip/protocols/BasicGossip.scala
package example.basicGossip.protocols

import peersim.core.Node
import scala.util.Random
import peersim.core.Network
import example.basicGossip.Usernode
import peersim.config.FastConfig
import peersim.core.Linkable
import peersim.transport.Transport
import example.basicGossip.Info
import peersim.edsim.EDProtocol
import peersim.cdsim.CDProtocol
import peersim.vector.SingleValueHolder
import peersim.config.Configuration

class BasicGossip(prefix: String) extends SingleValueHolder(prefix) with CDProtocol with EDProtocol {

  val networkSize = (Configuration.getInt("network.size", 1))
  val cycles = (Configuration.getInt("CYCLES", 1))
  val fanout = Math.log(networkSize).toInt
  var info: Int = 0

  override def nextCycle(node: Node, pid: Int): Unit = {
    
    node.getIndex match {
      case 0 => Network.get(0) match {
        case streamer: Usernode => sendInfo(streamer, fanout, pid, newInfo)
      }
      case _ => //only streamer does stuff at start of new cycle
    }
  }

  private def sendInfoAux(streamer: Usernode, info: Info, pid: Int) = {
    streamer.getProtocol(FastConfig.getLinkable(pid)) match {
      case link: Linkable if link.degree > 0 =>
        link.getNeighbor(Random.nextInt(link.degree)) match {
          case peern if peern.isUp => streamer.getProtocol(FastConfig.getTransport(pid)) match {
            case trans: Transport => trans.send(streamer, peern, info, pid)
            case _ => None
          }
          case peern if !peern.isUp => None
        }
      case _ => println("Could not send. Is degree > 0 ?")
    }

  }

  private def sendInfo(streamer: Usernode, fanout: Int, pid: Int, info: Info): Unit = {
    fanout match {
      case 0 =>
      case _ =>
        sendInfoAux(streamer, info, pid)
        sendInfo(streamer, fanout - 1, pid, info)
    }
  }


  private def newInfo: Info = {
    val currentInfo = info;
    info += 1;
    Network.get(0) match {
      case streamer: Usernode => Info(currentInfo, streamer, 0)
    }
  }

  def processEvent(node: Node, pid: Int, event: Object) {
    //for now, no 1 should send messages to streamer
  }

}

object BasicGossip extends BasicGossip("Basic Gossip")

